### DISCUSSION TITLE
Browser version for PKCS11

### DISCUSSION THREAD

**Discussion by AlainCNpt** [MatthiasValvekens/pyHanko#123] [Q&A]
Dear,

Has anyone tried (and succeeded) to combine PyHanko with a web-application?  There is a lot to be found regarding Python at the backend, which might be ok to sign documents that don't require front-end information like a PKCS11-Smart Card.

And there are also projects which also mention the use of the frontend, but I did not go deeply into those (as I don't know yet how to start if I would need to get this info to a PyHanko based script)

(I know Matthias is always ready to help, but I guess this will be for the community :))

BR
Alain

**Comment by MatthiasValvekens** [comment]
I'll assume that you're trying to find a way to talk to a client-side PKCS#11 device, not run pyHanko client-side (because even with PyScript, that's not going to work ;) ).

While I don't have a solution for you, here are some idle thoughts.

Essentially, you need two things:

 - A way to get the browser to sign stuff using a PKCS#11 device.
 - A way to communicate with a remote copy of pyHanko.

The second step is not too hard. You could implement a custom `Signer` on the pyHanko side and communicate asynchronously with the client through a web socket. The problem is mostly in getting the client to sign the data you want.

While I'm not as up-to-date on browser crypto as I perhaps should be, apparently the WebCrypto framework has some potentially relevant functionality. [Here's a JS library](https://github.com/PeculiarVentures/webcrypto-local) that looks potentially relevant. I don't know if it does what you want, and I've never used it myself, but I know some of the people involved, and they know their stuff.

If WebCrypto doesn't cut it, then perhaps a browser extension would work, depending on how much freedom the browser gives you to call external native software (!). You lose some interoperability and user-friendliness, though.

Security-wise this isn't an easy problem, though. The user has to trust your application not to feed it bogus hashes over the network, and since PKCS#11 modules run outside the browser, there's a security risk there as well---all things considered, I don't think there's a true "plug and play" solution here that works with zero config on the user's end. The security<>convenience tradeoff is significant.
